File: go/core/nvml_reader.go
// NVML library wrapper — Phase 2 GPU foundation.
//
// Mirrors python/src/dexcost/nvml_reader.py. The Go SDK avoids a hard
// dep on github.com/NVIDIA/go-nvml so the SDK builds + tests on GPU-less
// hosts; NVML access happens through a pluggable backend interface
// (NVMLBackend). The default backend is a noop that reports "no GPU".
//
// Consumers can register a real backend via SetNVMLBackend(); the GPU
// stack works on default-noop in unit tests with mocked devices.
//
// Decision #4: GetNVMLProductName applies NFC Unicode normalization +
// lowercase + whitespace collapse on the raw productName before returning
// — catalog alias matching depends on byte-level equality post-normalize
// because NVIDIA's productName ships with NBSP/NNBSP/ZWSP quirks across
// driver versions.
//
// Decision #8: GetNVMLProcessUtilization mutates a per-device
// lastSeenTimestamps map in place so callers persist NVML's sample
// buffer state across snapshot calls.
//
// Fail-silent contract (convention §9): every accessor returns nil (or
// false for booleans) on backend errors; the caller (typically
// GpuAccountant) decides the fallback policy.

package core

import (
	"strings"
	"sync"

	"golang.org/x/text/unicode/norm"
)

// NVMLProcessInfo is one compute-running-process record (PID + VRAM bytes).
type NVMLProcessInfo struct {
	PID           int
	UsedGPUMemory int64
}

// NVMLUtilSample is one process-utilization sample.
type NVMLUtilSample struct {
	PID       int
	SMUtil    int   // 0-100; percent of time SMs had ≥1 kernel running
	MemUtil   int   // 0-100; percent of time memory subsystem was busy
	TimeStamp int64 // microseconds since NVML epoch (monotonic)
}

// NVMLMemInfo is device-level memory totals.
type NVMLMemInfo struct {
	UsedBytes  int64
	TotalBytes int64
}

// NVMLBackend abstracts NVML so the SDK doesn't require pynvml/go-nvml at
// build time. The default backend returns "no GPU"; production deployments
// inject a real NVML-backed implementation; tests inject a deterministic
// mock.
type NVMLBackend interface {
	Available() bool
	Init() bool
	Shutdown()
	DeviceCount() (int, bool)
	ProductName(devIdx int) (string, bool)
	MIGMode(devIdx int) bool
	ComputeRunningProcesses(devIdx int) ([]NVMLProcessInfo, bool)
	// ProcessUtilization returns a list of utilization samples per PID
	// observed since `lastSeen[pid]`. Multi-sample-per-PID is required
	// for B2 integration math (Sprint 2 Theme C / §3.1.1, Go port);
	// pre-fix the API collapsed to a single sample and the accountant
	// could not integrate sm_util × dt across the task window.
	ProcessUtilization(devIdx int, lastSeen map[int]int64) (map[int][]NVMLUtilSample, bool)
	MemoryInfo(devIdx int) (NVMLMemInfo, bool)
}

// noopNVMLBackend is the default: reports "no GPU" for every call.
type noopNVMLBackend struct{}

func (noopNVMLBackend) Available() bool                { return false }
func (noopNVMLBackend) Init() bool                     { return false }
func (noopNVMLBackend) Shutdown()                      {}
func (noopNVMLBackend) DeviceCount() (int, bool)       { return 0, false }
func (noopNVMLBackend) ProductName(int) (string, bool) { return "", false }
func (noopNVMLBackend) MIGMode(int) bool               { return false }
func (noopNVMLBackend) ComputeRunningProcesses(int) ([]NVMLProcessInfo, bool) {
	return nil, false
}
func (noopNVMLBackend) ProcessUtilization(int, map[int]int64) (map[int][]NVMLUtilSample, bool) {
	return nil, false
}
func (noopNVMLBackend) MemoryInfo(int) (NVMLMemInfo, bool) { return NVMLMemInfo{}, false }

var (
	nvmlBackendMu sync.RWMutex
	nvmlBackend   NVMLBackend = noopNVMLBackend{}
)

// SetNVMLBackend swaps the global NVML backend. Used by consumers that
// vendor a real NVML library. nil reverts to the noop default.
func SetNVMLBackend(b NVMLBackend) {
	nvmlBackendMu.Lock()
	defer nvmlBackendMu.Unlock()
	if b == nil {
		nvmlBackend = noopNVMLBackend{}
		return
	}
	nvmlBackend = b
}

// SetNVMLBackendForTests is the test alias for SetNVMLBackend.
func SetNVMLBackendForTests(b NVMLBackend) { SetNVMLBackend(b) }

func getNVMLBackend() NVMLBackend {
	nvmlBackendMu.RLock()
	defer nvmlBackendMu.RUnlock()
	return nvmlBackend
}

// resetNVMLForTests reverts backend + warning state. Mirrors the
// _reset_warning_state helper from Python.
func resetNVMLForTests() {
	SetNVMLBackend(nil)
	nvmlWarnMu.Lock()
	nvmlWarned = map[string]struct{}{}
	nvmlWarnMu.Unlock()
}

// ResetNVMLForTests is the exported alias of resetNVMLForTests so other
// packages (adapters_test, etc.) can reset NVML state.
func ResetNVMLForTests() { resetNVMLForTests() }

// ─── log-once-per-failure-mode (convention §11) ─────────────────────────

var (
	nvmlWarnMu sync.Mutex
	nvmlWarned = map[string]struct{}{}
)

func nvmlWarnOnce(mode string) {
	nvmlWarnMu.Lock()
	defer nvmlWarnMu.Unlock()
	if _, seen := nvmlWarned[mode]; seen {
		return
	}
	nvmlWarned[mode] = struct{}{}
}

// ─── Decision #4 — NFC + lowercase + whitespace collapse ───────────────

// normalizeProductName applies NFC normalization, lowercases, and
// collapses any-and-all Unicode whitespace (including NBSP/NNBSP/ZWSP)
// into single ASCII spaces. Alias matching in the catalog depends on
// byte equality after this step.
func normalizeProductName(raw string) string {
	nfc := norm.NFC.String(raw)
	collapsed := strings.Join(strings.Fields(nfc), " ")
	return strings.ToLower(collapsed)
}

// ─── Public API ─────────────────────────────────────────────────────────

// NVMLAvailable reports whether NVML can be used at all.
func NVMLAvailable() bool {
	return getNVMLBackend().Available()
}

// InitNVML calls nvmlInit() if available. Returns true on success.
func InitNVML() bool {
	b := getNVMLBackend()
	if !b.Available() {
		nvmlWarnOnce("gpu_nvml_unavailable")
		return false
	}
	if !b.Init() {
		nvmlWarnOnce("gpu_nvml_init_failed")
		return false
	}
	return true
}

// ShutdownNVML best-effort shuts down NVML.
func ShutdownNVML() { getNVMLBackend().Shutdown() }

// GetNVMLDeviceCount returns the device count or nil on failure.
func GetNVMLDeviceCount() *int {
	b := getNVMLBackend()
	if !b.Available() {
		return nil
	}
	n, ok := b.DeviceCount()
	if !ok {
		nvmlWarnOnce("gpu_device_count_failed")
		return nil
	}
	return &n
}

// GetNVMLProductName returns the NFC-normalized, lowercased productName
// for device devIdx. nil on failure.
func GetNVMLProductName(devIdx int) *string {
	b := getNVMLBackend()
	if !b.Available() {
		return nil
	}
	raw, ok := b.ProductName(devIdx)
	if !ok {
		nvmlWarnOnce("gpu_product_name_failed")
		return nil
	}
	n := normalizeProductName(raw)
	return &n
}

// GetNVMLMIGMode reports whether MIG is currently enabled on device devIdx.
func GetNVMLMIGMode(devIdx int) bool {
	b := getNVMLBackend()
	if !b.Available() {
		return false
	}
	return b.MIGMode(devIdx)
}

// GetNVMLComputeRunningProcesses returns PIDs holding the GPU + VRAM usage.
// Returns nil on permission denied (load-bearing non-root-container case).
func GetNVMLComputeRunningProcesses(devIdx int) []NVMLProcessInfo {
	b := getNVMLBackend()
	if !b.Available() {
		return nil
	}
	out, ok := b.ComputeRunningProcesses(devIdx)
	if !ok {
		nvmlWarnOnce("gpu_nvml_permission_denied")
		return nil
	}
	return out
}

// GetNVMLProcessUtilization returns per-PID utilization samples for device
// devIdx and updates the lastSeen map in place (Decision #8 persistent
// state). Returns nil on failure.
//
// Returns map[pid][]NVMLUtilSample — multiple samples per PID, sorted
// by TimeStamp ascending. Sprint 2 Theme C / §3.1.1 (B2 Go port):
// pre-fix the wrapper collapsed samples to the latest per PID, losing
// the integration data the accountant now needs.
func GetNVMLProcessUtilization(devIdx int, lastSeen map[int]int64) map[int][]NVMLUtilSample {
	if lastSeen == nil {
		return nil
	}
	b := getNVMLBackend()
	if !b.Available() {
		return nil
	}
	out, ok := b.ProcessUtilization(devIdx, lastSeen)
	if !ok {
		nvmlWarnOnce("gpu_process_utilization_failed")
		return nil
	}
	// Record the max timestamp seen per PID — same as Python.
	for pid, samples := range out {
		for _, s := range samples {
			if s.TimeStamp > lastSeen[pid] {
				lastSeen[pid] = s.TimeStamp
			}
		}
	}
	return out
}

// GetNVMLMemoryInfo returns device-level VRAM info; nil on failure.
func GetNVMLMemoryInfo(devIdx int) *NVMLMemInfo {
	b := getNVMLBackend()
	if !b.Available() {
		return nil
	}
	m, ok := b.MemoryInfo(devIdx)
	if !ok {
		nvmlWarnOnce("gpu_memory_info_failed")
		return nil
	}
	return &m
}

// ─── Test backend ───────────────────────────────────────────────────────

// MockNVMLBackend is a deterministic backend for tests. The exported
// fields are configured before the backend is registered via
// SetNVMLBackendForTests; the interface methods are wired below.
type MockNVMLBackend struct {
	// Available toggles every accessor at once. Field is named with a
	// trailing underscore in struct layout to avoid clash with the
	// Available() method required by the NVMLBackend interface.
	Available      bool
	DeviceCount    int
	ProductNames   map[int]string
	MIGModes       map[int]bool
	Procs          map[int][]NVMLProcessInfo
	// Utilizations: map[devIdx][pid][]Sample. Single-element slice gives
	// the legacy single-sample behaviour; multi-element slices exercise
	// B2 integration math.
	Utilizations   map[int]map[int][]NVMLUtilSample
	Memory         map[int]NVMLMemInfo
	ProductNameErr map[int]bool
	DeviceCountErr bool

	// PerCallUtilization, when set, returns the slice element at index
	// invocationCount[devIdx] on each call, allowing tests to model NVML
	// snapshot evolution across start/end calls.
	PerCallUtilization map[int][]map[int][]NVMLUtilSample
	invocationCount    map[int]int

	mu sync.Mutex
}

// nvmlAvailable is the internal helper.
func (m *MockNVMLBackend) nvmlAvailable() bool {
	if m == nil {
		return false
	}
	return m.Available
}

// NVMLBackend interface methods. We use distinct method receiver names
// to avoid the field/method same-name collision.

func (m *MockNVMLBackend) Init() bool                    { return m.nvmlAvailable() }
func (m *MockNVMLBackend) Shutdown()                     {}

func (m *MockNVMLBackend) GetAvailable() bool { return m.nvmlAvailable() }

func (m *MockNVMLBackend) GetDeviceCount() (int, bool) {
	if !m.nvmlAvailable() || m.DeviceCountErr {
		return 0, false
	}
	return m.DeviceCount, true
}

func (m *MockNVMLBackend) GetProductName(devIdx int) (string, bool) {
	if !m.nvmlAvailable() {
		return "", false
	}
	if m.ProductNameErr != nil && m.ProductNameErr[devIdx] {
		return "", false
	}
	n, ok := m.ProductNames[devIdx]
	return n, ok
}

func (m *MockNVMLBackend) GetMIGMode(devIdx int) bool {
	if !m.nvmlAvailable() {
		return false
	}
	if m.MIGModes == nil {
		return false
	}
	return m.MIGModes[devIdx]
}

func (m *MockNVMLBackend) GetComputeRunningProcesses(devIdx int) ([]NVMLProcessInfo, bool) {
	if !m.nvmlAvailable() {
		return nil, false
	}
	if m.Procs == nil {
		return nil, true
	}
	return m.Procs[devIdx], true
}

func (m *MockNVMLBackend) GetProcessUtilization(devIdx int, _ map[int]int64) (map[int][]NVMLUtilSample, bool) {
	if !m.nvmlAvailable() {
		return nil, false
	}
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.PerCallUtilization != nil {
		if m.invocationCount == nil {
			m.invocationCount = map[int]int{}
		}
		seq := m.PerCallUtilization[devIdx]
		idx := m.invocationCount[devIdx]
		m.invocationCount[devIdx] = idx + 1
		if idx < len(seq) {
			return seq[idx], true
		}
		if len(seq) > 0 {
			return seq[len(seq)-1], true
		}
		return map[int][]NVMLUtilSample{}, true
	}
	if m.Utilizations == nil {
		return map[int][]NVMLUtilSample{}, true
	}
	return m.Utilizations[devIdx], true
}

func (m *MockNVMLBackend) GetMemoryInfo(devIdx int) (NVMLMemInfo, bool) {
	if !m.nvmlAvailable() {
		return NVMLMemInfo{}, false
	}
	if m.Memory == nil {
		return NVMLMemInfo{}, false
	}
	v, ok := m.Memory[devIdx]
	return v, ok
}

// mockAdapter wraps MockNVMLBackend to satisfy NVMLBackend (whose method
// names clash with the struct's exported field names).
type mockAdapter struct{ m *MockNVMLBackend }

func (a *mockAdapter) Available() bool               { return a.m.GetAvailable() }
func (a *mockAdapter) Init() bool                    { return a.m.Init() }
func (a *mockAdapter) Shutdown()                     { a.m.Shutdown() }
func (a *mockAdapter) DeviceCount() (int, bool)      { return a.m.GetDeviceCount() }
func (a *mockAdapter) ProductName(i int) (string, bool) { return a.m.GetProductName(i) }
func (a *mockAdapter) MIGMode(i int) bool            { return a.m.GetMIGMode(i) }
func (a *mockAdapter) ComputeRunningProcesses(i int) ([]NVMLProcessInfo, bool) {
	return a.m.GetComputeRunningProcesses(i)
}
func (a *mockAdapter) ProcessUtilization(i int, ls map[int]int64) (map[int][]NVMLUtilSample, bool) {
	return a.m.GetProcessUtilization(i, ls)
}
func (a *mockAdapter) MemoryInfo(i int) (NVMLMemInfo, bool) { return a.m.GetMemoryInfo(i) }

// AsBackend exposes the mock as an NVMLBackend (used by tests that
// register via SetNVMLBackendForTests).
func (m *MockNVMLBackend) AsBackend() NVMLBackend { return &mockAdapter{m: m} }

File: python/src/dexcost/nvml_reader.py
"""NVML library wrapper — Phase 2 GPU foundation.

Wraps NVIDIA's `pynvml` (aka `nvidia-ml-py`) library with the fail-silent
contract from convention §9. Every NVML call returns ``None`` (or ``False``
for boolean accessors) on missing driver / library / permission / device
errors rather than raising — the caller (typically :class:`GpuAccountant`)
decides the fallback policy per Decision #1's classification table.

``pynvml`` is an OPTIONAL dependency (install via ``pip install dexcost[gpu]``).
The top-level ``import pynvml`` is guarded so the SDK works on GPU-less
hosts and customers who haven't opted into the GPU extra.

Per Decision #4: ``get_product_name`` applies NFC Unicode normalization
+ lowercase + whitespace collapse on the raw NVML string before returning
— catalog alias matching depends on byte-level equality after normalization
because NVIDIA's productName carries non-breaking spaces and other Unicode
quirks across driver versions.

Per Decision #8: ``get_process_utilization`` takes a mutable
``last_seen_timestamps`` dict and updates it in place — the caller persists
per-PID timestamps across snapshot calls so NVML's sample buffer doesn't
silently lose intermediate samples.
"""

from __future__ import annotations

import logging
import threading
import unicodedata
from dataclasses import dataclass

_log = logging.getLogger(__name__)

try:
    import pynvml as _pynvml
    _NVML_AVAILABLE = True
except ImportError:
    _pynvml = None  # type: ignore[assignment]
    _NVML_AVAILABLE = False

# Module-level set of warning-mode tokens already logged in this process
# (convention §11 — log-once-per-failure-mode).
_warned_modes: set[str] = set()
_warn_lock = threading.Lock()


def _reset_warning_state() -> None:
    """Test-only: clear the warn-once tracking set."""
    with _warn_lock:
        _warned_modes.clear()


def _warn_once(mode: str, message: str) -> None:
    with _warn_lock:
        if mode in _warned_modes:
            return
        _warned_modes.add(mode)
    _log.warning(message)


# ─── Typed return values ─────────────────────────────────────────────────────


@dataclass(frozen=True)
class ProcessInfo:
    """One NVML compute-running-process record (PID + GPU memory usage)."""

    pid: int
    used_gpu_memory: int  # bytes; may be 0 when NVML reports NOT_AVAILABLE


@dataclass(frozen=True)
class UtilSample:
    """One NVML process-utilization sample for a single PID at one timestamp."""

    pid: int
    sm_util: int      # 0-100 — percent of time SMs had ≥1 kernel running
    mem_util: int     # 0-100 — percent of time memory subsystem was busy
    time_stamp: int   # microseconds since NVML epoch


@dataclass(frozen=True)
class MemInfo:
    """Device-level memory totals."""

    used_bytes: int
    total_bytes: int


# ─── Availability + init ─────────────────────────────────────────────────────


def nvml_available() -> bool:
    """True when ``pynvml`` was importable at SDK load time.

    Returning False here means ``nvidia-ml-py`` isn't installed (no
    ``pip install dexcost[gpu]``). NVML driver presence is a separate
    question — see ``init_nvml()``.
    """
    return _NVML_AVAILABLE and _pynvml is not None


def init_nvml() -> bool:
    """Call ``nvmlInit()``. Returns True on success, False on any failure.

    Fail modes (all silent + log-once):
    - pynvml not importable → ``gpu_pynvml_not_installed``
    - NVIDIA driver not loaded (NVML_ERROR_DRIVER_NOT_LOADED) → ``gpu_no_driver_in_container``
    - Library load failure (NVML_ERROR_LIBRARY_NOT_FOUND) → ``gpu_nvml_library_missing``
    """
    if not nvml_available():
        _warn_once(
            "gpu_pynvml_not_installed",
            "pynvml not installed; GPU capture disabled. "
            "Install with: pip install dexcost[gpu]",
        )
        return False
    try:
        _pynvml.nvmlInit()
        return True
    except _pynvml.NVMLError as exc:
        _warn_once(
            "gpu_nvml_init_failed",
            f"NVML init failed ({exc}); GPU capture disabled for this process",
        )
        return False


def shutdown_nvml() -> None:
    """Best-effort ``nvmlShutdown()``. Silent on any error."""
    if not nvml_available():
        return
    try:
        _pynvml.nvmlShutdown()
    except _pynvml.NVMLError:
        pass


# ─── Device enumeration ──────────────────────────────────────────────────────


def get_device_count() -> int | None:
    """Number of NVIDIA devices visible to NVML. ``None`` on failure."""
    if not nvml_available():
        return None
    try:
        return int(_pynvml.nvmlDeviceGetCount())
    except _pynvml.NVMLError as exc:
        _warn_once(
            "gpu_device_count_failed",
            f"nvmlDeviceGetCount failed ({exc})",
        )
        return None


def get_device_handle(index: int):
    """Opaque device handle (used by other accessors). ``None`` on failure."""
    if not nvml_available():
        return None
    try:
        return _pynvml.nvmlDeviceGetHandleByIndex(index)
    except _pynvml.NVMLError as exc:
        _warn_once(
            f"gpu_device_handle_failed:{index}",
            f"nvmlDeviceGetHandleByIndex({index}) failed ({exc})",
        )
        return None


# ─── Decision #4: NFC-normalized productName ─────────────────────────────────


def _normalize_product_name(raw: str) -> str:
    """NFC → lowercase → whitespace collapse (incl. NBSP / NNBSP / ZWSP)."""
    nfc = unicodedata.normalize("NFC", raw)
    # str.split() with no argument collapses ANY whitespace including
    # non-breaking space U+00A0, narrow no-break space U+202F, zero-width
    # characters that Python treats as whitespace, etc.
    return " ".join(nfc.split()).lower()


def get_product_name(handle) -> str | None:
    """Return the NVML productName, NFC-normalized + lowercased.

    Decision #4 — alias matching against the catalog depends on byte-level
    equality post-normalization.
    """
    if not nvml_available():
        return None
    try:
        raw = _pynvml.nvmlDeviceGetName(handle)
    except _pynvml.NVMLError as exc:
        _warn_once(
            "gpu_product_name_failed",
            f"nvmlDeviceGetName failed ({exc})",
        )
        return None
    if isinstance(raw, bytes):
        raw = raw.decode("utf-8", errors="replace")
    return _normalize_product_name(raw)


# ─── Per-PID compute processes (Decision #1 measurement-side primitive) ──────


def get_compute_running_processes(handle) -> list[ProcessInfo] | None:
    """List of PIDs holding the GPU + their VRAM usage.

    Returns ``None`` on ``NVML_ERROR_NO_PERMISSION`` — the load-bearing
    non-root-container case from Decision #1. The caller's cgroup walk
    then degrades to self-PID-only.
    """
    if not nvml_available():
        return None
    try:
        raw = _pynvml.nvmlDeviceGetComputeRunningProcesses(handle)
    except _pynvml.NVMLError as exc:
        _warn_once(
            "gpu_nvml_permission_denied",
            f"nvmlDeviceGetComputeRunningProcesses failed ({exc}); "
            "GpuAccountant will degrade to self-PID-only",
        )
        return None
    out: list[ProcessInfo] = []
    for p in raw:
        try:
            mem = int(getattr(p, "usedGpuMemory", 0) or 0)
        except (TypeError, ValueError):
            mem = 0
        out.append(ProcessInfo(pid=int(p.pid), used_gpu_memory=mem))
    return out


# ─── Per-PID utilization with persistent timestamp state (Decision #8) ───────


def get_process_utilization(
    handle, last_seen_timestamps: dict[int, int],
) -> dict[int, list[UtilSample]] | None:
    """Return per-PID utilization samples; UPDATE timestamps dict in place.

    Decision #8 — NVML's ``nvmlDeviceGetProcessUtilization`` returns samples
    accumulated since ``lastSeenTimeStamp``. The caller persists per-PID
    timestamps across snapshot calls so we don't lose samples between calls
    when the internal NVML buffer wraps.

    Returns ``dict[int, list[UtilSample]]`` — multiple samples per PID,
    sorted by ``time_stamp`` ascending. Pre-B2 the wrapper collapsed
    samples to the latest per PID, throwing away the integration data
    the accountant now needs (Sprint 2 Theme C / §3.1.1).

    The earliest baseline call passes ``last_seen_timestamps={}``; each
    subsequent call passes the dict the previous call updated.
    """
    if not nvml_available():
        return None
    # Use the minimum timestamp across known PIDs (or 0 for first call) to
    # capture samples for any new PIDs that joined since last snapshot.
    base_ts = min(last_seen_timestamps.values(), default=0) if last_seen_timestamps else 0
    try:
        raw = _pynvml.nvmlDeviceGetProcessUtilization(handle, base_ts)
    except _pynvml.NVMLError as exc:
        _warn_once(
            "gpu_process_utilization_failed",
            f"nvmlDeviceGetProcessUtilization failed ({exc})",
        )
        return None
    out: dict[int, list[UtilSample]] = {}
    for s in raw:
        pid = int(s.pid)
        ts = int(getattr(s, "timeStamp", 0) or 0)
        sm = int(getattr(s, "smUtil", 0) or 0)
        mem = int(getattr(s, "memUtil", 0) or 0)
        out.setdefault(pid, []).append(
            UtilSample(pid=pid, sm_util=sm, mem_util=mem, time_stamp=ts)
        )
        # last_seen tracks the MAX ts per PID across this batch.
        if ts > last_seen_timestamps.get(pid, 0):
            last_seen_timestamps[pid] = ts
    # Sort each PID's samples by timestamp ascending — the integration
    # in the accountant assumes monotonic ordering.
    for pid in out:
        out[pid].sort(key=lambda s: s.time_stamp)
    return out


# ─── Memory + MIG ────────────────────────────────────────────────────────────


def get_memory_info(handle) -> MemInfo | None:
    """Device-level used / total VRAM bytes. ``None`` on failure."""
    if not nvml_available():
        return None
    try:
        info = _pynvml.nvmlDeviceGetMemoryInfo(handle)
    except _pynvml.NVMLError as exc:
        _warn_once(
            "gpu_memory_info_failed",
            f"nvmlDeviceGetMemoryInfo failed ({exc})",
        )
        return None
    return MemInfo(used_bytes=int(info.used), total_bytes=int(info.total))


def get_mig_mode(handle) -> bool:
    """True when MIG is enabled on this device (Decision #2 detection)."""
    if not nvml_available():
        return False
    try:
        current, _pending = _pynvml.nvmlDeviceGetMigMode(handle)
    except _pynvml.NVMLError:
        # Older GPUs without MIG support → fail-silent (NOT an error)
        return False
    enable_const = getattr(_pynvml, "NVML_DEVICE_MIG_ENABLE", 1)
    return current == enable_const
